lib/Target/AVR/AVRInstrInfo.cpp

```diff
@@ -45,6 +45,30 @@ extern bool RA_InSpillerCode;
 AVRInstrInfo::AVRInstrInfo()
     : AVRGenInstrInfo(AVR::ADJCALLSTACKDOWN, AVR::ADJCALLSTACKUP), RI() {}
 
+// FIXME: BEFORE MERGING: Is there a cleaner tablegen method of doing this?
```

Comment:
> I think I remembered the calling convention notes saying that r0 should generally be 0 but it doesn't look that way.

Yeah, that code is in [`AVRTargetLowering::insertMul`](https://github.com/avr-llvm/llvm/blob/fb6e1cecad6c2f26be9e441ee4c8c51518545895/lib/Target/AVR/AVRISelLowering.cpp#L1485-L1499), and it unconditionally inserts the clear to zero. I don't think that the *rest* of LLVM knows that `r1` (not `r0`) is supposed to contain zero, so it always loads an immediate zero elsewhere.

And in *this* case, we don't need the zero for any other work in this method, so we could have left the value in place and cleared it to zero before we returned.



> Did you figure out the weird movs?

Yes, those were the result of the "move 16-bit register to 8-bit register" strangeness and are now gone.